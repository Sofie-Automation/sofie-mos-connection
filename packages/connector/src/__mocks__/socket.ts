/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Socket } from 'net'

import * as iconv from 'iconv-lite'
import { getMessageId } from '../__tests__/lib.js'
import { getMosTypes } from '@mos-connection/model'
import { Mocked, vitest } from 'vitest'
iconv.encodingExists('utf16-be')

const mosTypes = getMosTypes(true)

// Mock the Socket class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<ISocketMock> = []

export class SocketMock extends EventEmitter implements Socket {
	// Mock implementation:
	bufferSize = 0
	remoteAddress = ''
	remoteFamily = ''
	remotePort = 0
	localAddress = ''
	localPort = 0
	bytesRead = 0
	bytesWritten = 0
	connecting = false
	destroyed = false
	writable = false
	readable = false

	readonly readableEncoding: BufferEncoding | null = null
	readonly readableEnded: boolean = false
	readonly readableFlowing: boolean | null = null

	readonly writableEnded: boolean = false
	readonly writableFinished: boolean = false
	readonly writableHighWaterMark: number = 0
	readonly writableLength: number = 0
	readonly writableObjectMode: boolean = false

	readonly readableHighWaterMark: number = 0
	readonly readableLength: number = 0
	readonly readableObjectMode: boolean = false

	public name = ''
	public connectedPort = 0
	public connectedHost = ''

	// [Symbol.asyncIterator](): AsyncIterableIterator<any>

	private _responses: Array<ReplyTypes> = []
	private _autoReplyToHeartBeat = true
	public mockConnectCount = 0

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		// Wrap member functions in mocks:
		this.connect = vitest.fn(this.connect)
		this.setEncoding = vitest.fn(this.setEncoding)
		this.destroy = vitest.fn(this.destroy)
		this.pause = vitest.fn(this.pause)
		this.resume = vitest.fn(this.resume)
		this.setTimeout = vitest.fn(this.setTimeout)
		this.setNoDelay = vitest.fn(this.setNoDelay)
		this.setKeepAlive = vitest.fn(this.setKeepAlive)
		this.address = vitest.fn(this.address)
		this.unref = vitest.fn(this.unref)
		this.ref = vitest.fn(this.ref)
		this.end = vitest.fn(this.end)
		this._write = vitest.fn(this._write)
		this.setDefaultEncoding = vitest.fn(this.setDefaultEncoding)
		this._read = vitest.fn(this._read)
		this.read = vitest.fn(this.read)
		this.isPaused = vitest.fn(this.isPaused)
		this.pipe = vitest.fn(this.pipe) as any
		this.unpipe = vitest.fn(this.unpipe)
		this.unshift = vitest.fn(this.unshift)
		this.wrap = vitest.fn(this.wrap)
		this.push = vitest.fn(this.push)
		this._destroy = vitest.fn(this._destroy)
		this._final = vitest.fn(this._final)
		this.cork = vitest.fn(this.cork)
		this.uncork = vitest.fn(this.uncork)

		this.mockSentMessage0 = vitest.fn(this.mockSentMessage0)
		this.mockSentMessage = vitest.fn(this.mockSentMessage)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): ISocketMock[] {
		return instances
	}

	write(data: unknown, encoding: unknown): boolean {
		this.mockSentMessage0(data, encoding as any)
		return true
	}
	// @ts-expect-error mock
	connect(port: number, host: string): this {
		this.mockConnectCount++
		this.connectedPort = port
		this.connectedHost = host

		return this
	}
	setEncoding(): this {
		return this
	}
	destroy(_error?: Error): this {
		this.destroyed = true
		this.emit('close')
		/* nothing */
		return this
	}
	pause(): this {
		return this
	}
	resume(): this {
		return this
	}
	setTimeout(timeout: number, callback?: (...args: any[]) => void): this {
		if (callback) setTimeout(callback, timeout)
		return this
	}
	setNoDelay(_noDelay?: boolean): this {
		return this
	}
	setKeepAlive(_enable?: boolean, _initialDelay?: number): this {
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 100, family: 'localhost', address: '127.0.0.1' }
	}
	unref(): this {
		return this
	}
	ref(): this {
		return this
	}
	end(): this {
		return this
	}
	_write(_chunk: unknown, _encoding: string, callback: () => void): void {
		callback() /* nothing */
	}
	setDefaultEncoding(_encoding: string): this {
		return this
	}
	_read(_size: number): void {
		// nothing
	}
	read(_size?: number): void {
		// nothing
	}
	isPaused(): boolean {
		return false
	}
	pipe<T extends NodeJS.WritableStream>(destination: T, _options?: { end?: boolean }): T {
		return destination
	}
	unpipe(_destination: NodeJS.WritableStream): this {
		return this
	}
	unshift(_chunk: unknown): void {
		// nothing
	}
	wrap(_oldStream: NodeJS.ReadableStream): this {
		return this
	}
	push(_chunk: unknown, _encoding?: string): boolean {
		return true
	}
	_destroy(): void {
		/* nothing */
	}
	_final(): void {
		/* nothing */
	}
	cork(): void {
		/* nothing */
	}
	uncork(): void {
		/* nothing */
	}

	// ------------------------------------------------------------------------
	// Mock methods:
	mockEmitConnected(): void {
		if (this.connectedPort === 10542) {
			// don't reply on heartbeats on query port
			this._autoReplyToHeartBeat = false
		}

		this.emit('connect')
	}
	mockEmitClose(): void {
		this.emit('close')
	}

	mockSentMessage0(data: unknown, encoding: string): void {
		if (this._autoReplyToHeartBeat) {
			const str: string = typeof data === 'string' ? data : this.decode(data as any)

			if (/<heartbeat>/.exec(str)) {
				try {
					const mosID = (str.match(/<mosID>([^<]+)<\/mosID>/) || [])[1]
					const ncsID = (str.match(/<ncsID>([^<]+)<\/ncsID>/) || [])[1]
					const messageId = getMessageId(str)

					const repl = `<mos>
<mosID>${mosID}</mosID>
<ncsID>${ncsID}</ncsID>
<messageID>${messageId}</messageID>\
  <heartbeat>
    <time>${mosTypes.mosTime.stringify(mosTypes.mosTime.create(Date.now()))}</time>
  </heartbeat>
</mos>\r\n`
					this.mockReceiveMessage(this.encode(repl))
				} catch (e) {
					// eslint-disable-next-line no-console
					console.error('mockReply', str)
					throw e
				}
				return
			}
		}
		this.mockSentMessage(data, encoding)
	}
	mockSentMessage(data: unknown, _encoding: string): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			setTimeoutOrg(() => {
				Promise.resolve(typeof cb === 'function' ? cb(data) : cb)
					.then((msg) => {
						if (msg) {
							if (Array.isArray(msg)) {
								for (const m of msg) {
									this.mockReceiveMessage(m)
								}
							} else {
								this.mockReceiveMessage(msg)
							}
						}
					})
					// eslint-disable-next-line no-console
					.catch(console.error)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: ReplyTypes): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		void this._responses.splice(0, 9999)
		// @ts-expect-error mock hack
		this.mockSentMessage0['mockClear']()
		// @ts-expect-error mock hack
		this.mockSentMessage['mockClear']()
	}
	async mockWaitForSentMessages(): Promise<void> {
		return new Promise<void>((resolve) => {
			const check = () => {
				if (this._responses.length === 0) {
					resolve()
				} else {
					setTimeoutOrg(() => {
						check()
					}, 1)
				}
			}
			check()
		})
	}

	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
	setAutoReplyToHeartBeat(autoReplyToHeartBeat: boolean): void {
		this._autoReplyToHeartBeat = autoReplyToHeartBeat
	}
	get autoReplyToHeartBeat(): boolean {
		return this._autoReplyToHeartBeat
	}
}

type SimpleTypes = string | string[] | false | Buffer | Buffer[]

type ReplyTypes =
	| SimpleTypes
	| ((data: any) => SimpleTypes)
	| Promise<SimpleTypes>
	| ((data: any) => Promise<SimpleTypes>)

export type ISocketMock = Mocked<SocketMock>
