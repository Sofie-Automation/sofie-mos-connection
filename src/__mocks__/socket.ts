/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Socket } from 'net'
import { MosTime } from '../dataTypes/mosTime'

import * as iconv from 'iconv-lite'
import { getMessageId } from '../__tests__/lib'
iconv.encodingExists('utf16-be')

// Mock the Socket class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<ISocketMock> = []

export class SocketMock extends EventEmitter implements Socket {
	// Mock implementation:
	bufferSize = 0
	remoteAddress = ''
	remoteFamily = ''
	remotePort = 0
	localAddress = ''
	localPort = 0
	bytesRead = 0
	bytesWritten = 0
	connecting = false
	destroyed = false
	writable = false
	readable = false

	readonly readableEncoding: BufferEncoding | null = null
	readonly readableEnded: boolean = false
	readonly readableFlowing: boolean | null = null

	readonly writableEnded: boolean = false
	readonly writableFinished: boolean = false
	readonly writableHighWaterMark: number = 0
	readonly writableLength: number = 0
	readonly writableObjectMode: boolean = false

	readonly readableHighWaterMark: number = 0
	readonly readableLength: number = 0
	readonly readableObjectMode: boolean = false

	public name = ''
	public connectedPort = 0
	public connectedHost = ''

	// [Symbol.asyncIterator](): AsyncIterableIterator<any>

	private _responses: Array<ReplyTypes> = []
	private _autoReplyToHeartBeat = true

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		// Wrap member functions in mocks:
		this.connect = jest.fn(this.connect)
		this.setEncoding = jest.fn(this.setEncoding)
		this.destroy = jest.fn(this.destroy)
		this.pause = jest.fn(this.pause)
		this.resume = jest.fn(this.resume)
		this.setTimeout = jest.fn(this.setTimeout)
		this.setNoDelay = jest.fn(this.setNoDelay)
		this.setKeepAlive = jest.fn(this.setKeepAlive)
		this.address = jest.fn(this.address)
		this.unref = jest.fn(this.unref)
		this.ref = jest.fn(this.ref)
		this.end = jest.fn(this.end)
		this._write = jest.fn(this._write)
		this.setDefaultEncoding = jest.fn(this.setDefaultEncoding)
		this._read = jest.fn(this._read)
		this.read = jest.fn(this.read)
		this.isPaused = jest.fn(this.isPaused)
		this.pipe = jest.fn(this.pipe)
		this.unpipe = jest.fn(this.unpipe)
		this.unshift = jest.fn(this.unshift)
		this.wrap = jest.fn(this.wrap)
		this.push = jest.fn(this.push)
		this._destroy = jest.fn(this._destroy)
		this._final = jest.fn(this._final)
		this.cork = jest.fn(this.cork)
		this.uncork = jest.fn(this.uncork)

		this.mockSentMessage0 = jest.fn(this.mockSentMessage0)
		this.mockSentMessage = jest.fn(this.mockSentMessage)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): ISocketMock[] {
		return instances
	}

	write(data: unknown, encoding: unknown): boolean {
		this.mockSentMessage0(data, encoding as any)
		return true
	}
	// @ts-expect-error mock
	connect(port: number, host: string): SocketMock {
		this.connectedPort = port
		this.connectedHost = host

		if (this.connectedPort === 10542) {
			// don't reply on heartbeats on query port
			this._autoReplyToHeartBeat = false
		}

		this.emit('connect')
		return this
	}
	setEncoding(): this {
		return this
	}
	destroy(_error?: Error): this {
		this.destroyed = true
		this.emit('close')
		/* nothing */
		return this
	}
	pause(): this {
		return this
	}
	resume(): this {
		return this
	}
	setTimeout(timeout: number, callback?: (...args: any[]) => void): this {
		if (callback) setTimeout(callback, timeout)
		return this
	}
	setNoDelay(_noDelay?: boolean): this {
		return this
	}
	setKeepAlive(_enable?: boolean, _initialDelay?: number): this {
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 100, family: 'localhost', address: '127.0.0.1' }
	}
	unref(): this {
		return this
	}
	ref(): this {
		return this
	}
	end(): this {
		return this
	}
	_write(_chunk: unknown, _encoding: string, callback: () => void): void {
		callback() /* nothing */
	}
	setDefaultEncoding(_encoding: string): this {
		return this
	}
	_read(_size: number): void {
		// nothing
	}
	read(_size?: number): void {
		// nothing
	}
	isPaused(): boolean {
		return false
	}
	pipe<T extends NodeJS.WritableStream>(destination: T, _options?: { end?: boolean }): T {
		return destination
	}
	unpipe(_destination: NodeJS.WritableStream): this {
		return this
	}
	unshift(_chunk: unknown): void {
		// nothing
	}
	wrap(_oldStream: NodeJS.ReadableStream): this {
		return this
	}
	push(_chunk: unknown, _encoding?: string): boolean {
		return true
	}
	_destroy(): void {
		/* nothing */
	}
	_final(): void {
		/* nothing */
	}
	cork(): void {
		/* nothing */
	}
	uncork(): void {
		/* nothing */
	}

	// ------------------------------------------------------------------------
	// Mock methods:
	mockSentMessage0(data: unknown, encoding: string): void {
		if (this._autoReplyToHeartBeat) {
			const str: string = typeof data === 'string' ? data : this.decode(data as any)

			if (str.match(/<heartbeat>/)) {
				try {
					const mosID = (str.match(/<mosID>([^<]+)<\/mosID>/) || [])[1]
					const ncsID = (str.match(/<ncsID>([^<]+)<\/ncsID>/) || [])[1]
					const messageId = getMessageId(str)

					const repl = `<mos>
<mosID>${mosID}</mosID>
<ncsID>${ncsID}</ncsID>
<messageID>${messageId}</messageID>\
  <heartbeat>
    <time>${new MosTime().toString()}</time>
  </heartbeat>
</mos>\r\n`
					this.mockReceiveMessage(this.encode(repl))
				} catch (e) {
					console.error('mockReply', str)
					throw e
				}
				return
			}
		}
		this.mockSentMessage(data, encoding)
	}
	mockSentMessage(data: unknown, _encoding: string): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			setTimeoutOrg(() => {
				Promise.resolve(typeof cb === 'function' ? cb(data) : cb)
					.then((msg) => {
						if (msg) {
							if (Array.isArray(msg)) {
								for (const m of msg) {
									this.mockReceiveMessage(m)
								}
							} else {
								this.mockReceiveMessage(msg)
							}
						}
					})
					.catch(console.error)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: ReplyTypes): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		this._responses.splice(0, 9999)
		// @ts-expect-error mock hack
		this.mockSentMessage0['mockClear']()
		// @ts-expect-error mock hack
		this.mockSentMessage['mockClear']()
	}
	async mockWaitForSentMessages(): Promise<void> {
		return new Promise<void>((resolve) => {
			const check = () => {
				if (this._responses.length === 0) {
					resolve()
				} else {
					setTimeoutOrg(() => {
						check()
					}, 1)
				}
			}
			check()
		})
	}

	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
	setAutoReplyToHeartBeat(autoReplyToHeartBeat: boolean): void {
		this._autoReplyToHeartBeat = autoReplyToHeartBeat
	}
}

type SimpleTypes = string | string[] | false | Buffer | Buffer[]

type ReplyTypes =
	| SimpleTypes
	| ((data: any) => SimpleTypes)
	| Promise<SimpleTypes>
	| ((data: any) => Promise<SimpleTypes>)

export type ISocketMock = jest.Mocked<SocketMock>
