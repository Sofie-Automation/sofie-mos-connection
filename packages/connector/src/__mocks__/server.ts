/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Server } from 'net'
import { SocketMock } from './socket'

import * as iconv from 'iconv-lite'
iconv.encodingExists('utf16-be')

// Mock the Server class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<IServerMock> = []

export class ServerMock extends EventEmitter implements Server {
	// Mock implementation:

	maxConnections = 0
	connections = 0
	listening = false
	listenToPort = 0

	private _responses: Array<(data: any) => string | Buffer> = []

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		this.listen = jest.fn(this.listen)
		this.close = jest.fn(this.close)
		this.address = jest.fn(this.address)
		this.getConnections = jest.fn(this.getConnections)
		this.ref = jest.fn(this.ref)
		this.unref = jest.fn(this.unref)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): IServerMock[] {
		return instances
	}

	// @ts-expect-error mock hack
	listen(port: number): this {
		this.listenToPort = port
		setTimeoutOrg(() => {
			this.listening = true
			this.emit('listening', {})
		}, 1)
		return this
	}
	close(callback?: () => void): this {
		if (callback) callback()
		setTimeoutOrg(() => {
			this.emit('close')
		}, 1)
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 0, family: 'string', address: 'string' }
	}
	getConnections(cb: (error: Error | null, count: number) => void): this {
		cb(null, 0)
		return this
	}
	ref(): this {
		return this
	}
	unref(): this {
		return this
	}

	// Mock methods:
	mockNewConnection(): SocketMock {
		// "Someone has connected"
		const socket = new SocketMock()
		socket.on('data', (d) => {
			this.emit('data', d)
		})

		this.emit('connection', socket)

		return socket
	}

	mockSentMessage(data: string, _encoding: unknown): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			let msg

			setTimeoutOrg(() => {
				if (typeof cb === 'string') {
					msg = cb
				} else {
					msg = cb(data)
				}

				this.mockReceiveMessage(msg)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: (data: any) => string | Buffer): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		this._responses.splice(0, 9999)
	}
	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
}

export type IServerMock = jest.Mocked<ServerMock>
