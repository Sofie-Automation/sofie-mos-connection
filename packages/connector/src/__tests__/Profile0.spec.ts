import {
	checkMessageSnapshot,
	clearMocks,
	decode,
	DEFAULT_TIMEOUT,
	delay,
	doBeforeAll,
	encode,
	fakeIncomingMessage,
	getMessageId,
	getMosConnection,
	getMosDevice,
	getXMLReply,
	mosTypes,
	setupMocks,
} from './lib'
import { MosConnection, MosDevice, IMOSObject, IMOSListMachInfo } from '..'
import { SocketMock } from '../__mocks__/socket'
import { xmlData, xmlApiData } from '../__mocks__/testData'

/* eslint-disable @typescript-eslint/no-unused-vars */
// @ts-ignore imports are unused
import { Socket } from 'net'
/* eslint-enable @typescript-eslint/no-unused-vars */

beforeAll(() => {
	setupMocks()
})
beforeEach(() => {
	clearMocks()
})
describe('Profile 0', () => {
	let mosDevice: MosDevice
	let mosConnection: MosConnection

	let socketMockLower: SocketMock
	let socketMockUpper: SocketMock
	let socketMockQuery: SocketMock

	let serverSocketMockLower: SocketMock
	let serverSocketMockUpper: SocketMock
	let serverSocketMockQuery: SocketMock

	let onRequestMachineInfo: jest.Mock<any, any>
	let onRequestMOSObject: jest.Mock<any, any>
	let onRequestAllMOSObjects: jest.Mock<any, any>

	beforeAll(async () => {
		mosConnection = await getMosConnection(
			{
				'0': true,
				'1': true, // Must support at least one other profile
			},
			true
		)
		mosDevice = await getMosDevice(mosConnection)

		// Profile 0:
		onRequestMachineInfo = jest.fn(async () => {
			return xmlApiData.machineInfo
		})
		mosDevice.onRequestMachineInfo(async (): Promise<IMOSListMachInfo> => {
			return onRequestMachineInfo()
		})
		// Profile 1:
		onRequestMOSObject = jest.fn(async () => {
			return xmlApiData.mosObj
		})
		onRequestAllMOSObjects = jest.fn(async () => {
			return [xmlApiData.mosObj, xmlApiData.mosObj2]
		})
		mosDevice.onRequestMOSObject(async (objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})
		mosDevice.onRequestAllMOSObjects(async (): Promise<Array<IMOSObject>> => {
			return onRequestAllMOSObjects()
		})
		const b = doBeforeAll()
		socketMockLower = b.socketMockLower
		socketMockUpper = b.socketMockUpper
		socketMockQuery = b.socketMockQuery
		serverSocketMockLower = b.serverSocketMockLower
		serverSocketMockUpper = b.serverSocketMockUpper
		serverSocketMockQuery = b.serverSocketMockQuery

		mosDevice.checkProfileValidness()
		mosConnection.checkProfileValidness()
	})
	afterAll(async () => {
		await mosDevice.dispose()
		await mosConnection.dispose()
	})
	beforeEach(() => {
		onRequestMOSObject.mockClear()
		onRequestAllMOSObjects.mockClear()

		serverSocketMockLower.mockClear()
		serverSocketMockUpper.mockClear()
		if (serverSocketMockQuery) serverSocketMockQuery.mockClear()
		socketMockLower.mockClear()
		socketMockUpper.mockClear()
		if (socketMockQuery) socketMockQuery.mockClear()
	})
	test('init', async () => {
		expect(mosDevice).toBeTruthy()
		expect(socketMockLower).toBeTruthy()
		expect(socketMockUpper).toBeTruthy()
		expect(serverSocketMockLower).toBeTruthy()
	})
	test('heartbeat from other party', async () => {
		expect(serverSocketMockLower).toBeTruthy()

		serverSocketMockLower.setAutoReplyToHeartBeat(false) // Handle heartbeat manually

		const serverReply: jest.Mock<any, any> = jest.fn(() => false)
		serverSocketMockLower.mockAddReply(serverReply)
		// Fake incoming message on socket:
		const sendMessageId = await fakeIncomingMessage(serverSocketMockLower, xmlData.heartbeat)
		await delay(10) // to allow for async timers & events to triggered

		expect(serverReply).toHaveBeenCalledTimes(1)

		const msg = serverSocketMockLower.decode(serverReply.mock.calls[0][0])

		expect(msg).toMatch(/<heartbeat/)
		expect(msg).toMatch('<messageID>' + sendMessageId)
	})
	test('send heartbeats', async () => {
		socketMockLower.setAutoReplyToHeartBeat(false) // Handle heartbeat manually
		socketMockUpper.setAutoReplyToHeartBeat(false) // Handle heartbeat manually

		const heartbeatCount = {
			upper: 0,
			lower: 0,
		}

		const mockReply = (portType: 'upper' | 'lower') => {
			return (data: any) => {
				const str = decode(data)
				const messageID = getMessageId(str)

				if (str.match(/<heartbeat/)) {
					const repl = getXMLReply(messageID, xmlData.heartbeat)
					heartbeatCount[portType]++
					return encode(repl)
				} else throw new Error('Mock: Unhandled message: ' + str)
			}
		}

		for (let i = 0; i < 100; i++) {
			socketMockUpper.mockAddReply(mockReply('upper'))
			socketMockLower.mockAddReply(mockReply('lower'))
		}

		// During this time, there should have been sent a few heartbeats to the server:
		await delay(DEFAULT_TIMEOUT * 4.5)
		expect(heartbeatCount.upper).toBeGreaterThanOrEqual(4)
		expect(heartbeatCount.lower).toBeGreaterThanOrEqual(4)
		expect(mosDevice.getConnectionStatus()).toMatchObject({ PrimaryConnected: true })
	})

	test('unknown party connects', async () => {
		expect(serverSocketMockLower).toBeTruthy()
		serverSocketMockLower.setAutoReplyToHeartBeat(false)
		const serverReply: jest.Mock<any, any> = jest.fn(() => false)
		serverSocketMockLower.mockAddReply(serverReply)

		// Fake incoming message on socket:
		let sendMessageId = await fakeIncomingMessage(serverSocketMockLower, xmlData.heartbeat, 'ourUnknownMosId')
		await delay(10) // to allow for async timers & events to triggered

		expect(serverReply).toHaveBeenCalledTimes(1)
		let msg = serverSocketMockLower.decode(serverReply.mock.calls[0][0])
		expect(msg).toMatch(/<mosAck>/)
		expect(msg).toMatch('<messageID>' + sendMessageId)
		expect(msg).toMatch('<status>NACK')

		serverReply.mockClear()
		serverSocketMockLower.mockAddReply(serverReply)

		// Fake incoming message on socket:
		sendMessageId = await fakeIncomingMessage(
			serverSocketMockLower,
			xmlData.heartbeat,
			undefined,
			'theirUnknownMosId'
		)
		await delay(10) // to allow for async timers & events to triggered

		expect(serverReply).toHaveBeenCalledTimes(1)
		msg = serverSocketMockLower.decode(serverReply.mock.calls[0][0])
		expect(msg).toMatch(/<mosAck>/)
		expect(msg).toMatch('<messageID>' + sendMessageId)
		expect(msg).toMatch('<status>NACK')
	})
	// TODO: listMachInfo

	test('requestMachineInfo', async () => {
		// Prepare mock server response:
		const mockReply = jest.fn((data) => {
			const str = decode(data)
			const messageID = getMessageId(str)
			const repl = getXMLReply(messageID, xmlData.machineInfo)
			return encode(repl)
		})
		socketMockLower.mockAddReply(mockReply)
		if (!xmlApiData.mosObj.ID) throw new Error('xmlApiData.mosObj.ID not set')

		const returnedMachineInfo: IMOSListMachInfo = await mosDevice.requestMachineInfo()
		expect(mockReply).toHaveBeenCalledTimes(1)
		const msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<reqMachInfo\/>/)
		checkMessageSnapshot(msg)

		expect(returnedMachineInfo).toMatchObject({
			manufacturer: mosTypes.mosString128.create('RadioVision, Ltd.'),
			model: mosTypes.mosString128.create('TCS6000'),
			hwRev: mosTypes.mosString128.create(''),
			swRev: mosTypes.mosString128.create('2.1.0.37'),
			DOM: mosTypes.mosString128.create(''),
			SN: mosTypes.mosString128.create('927748927'),
			ID: mosTypes.mosString128.create('airchache.newscenter.com'),
			time: mosTypes.mosTime.create('2009-04-11T17:20:42'),
			opTime: mosTypes.mosTime.create('2009-03-01T23:55:10'),
			mosRev: mosTypes.mosString128.create('2.8.2'),

			supportedProfiles: {
				deviceType: 'NCS',
				profile0: true,
				profile1: true,
				profile2: true,
				profile3: true,
				profile4: true,
				profile5: true,
				profile6: true,
				profile7: true,
			},
		} as IMOSListMachInfo)
		expect(returnedMachineInfo).toMatchSnapshot()
	})
})
