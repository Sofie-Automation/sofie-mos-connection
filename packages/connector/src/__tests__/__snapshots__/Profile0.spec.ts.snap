// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile 0 requestMachineInfo - empty <opTime> 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <reqMachInfo/>
</mos>"
`;

exports[`Profile 0 requestMachineInfo - missing <opTime> 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <reqMachInfo/>
</mos>"
`;

exports[`Profile 0 requestMachineInfo 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <reqMachInfo/>
</mos>"
`;

exports[`Profile 0 requestMachineInfo 2`] = `
{
  "DOM": {
    "_mosString128": "",
  },
  "ID": {
    "_mosString128": "airchache.newscenter.com",
  },
  "SN": {
    "_mosString128": "927748927",
  },
  "hwRev": {
    "_mosString128": "",
  },
  "manufacturer": {
    "_mosString128": "RadioVision, Ltd.",
  },
  "model": {
    "_mosString128": "TCS6000",
  },
  "mosRev": {
    "_mosString128": "2.8.2",
  },
  "opTime": {
    "_mosTime": 1235951710000,
    "_timezone": "Z",
    "_timezoneOffset": 0,
  },
  "supportedProfiles": {
    "deviceType": "NCS",
    "profile0": true,
    "profile1": true,
    "profile2": true,
    "profile3": true,
    "profile4": true,
    "profile5": true,
    "profile6": true,
    "profile7": true,
  },
  "swRev": {
    "_mosString128": "2.1.0.37",
  },
  "time": {
    "_mosTime": 1239470442000,
    "_timezone": "Z",
    "_timezoneOffset": 0,
  },
}
`;
